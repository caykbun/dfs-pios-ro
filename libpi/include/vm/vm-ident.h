#ifndef __VM_IDENT_H__
#define __VM_IDENT_H__

#include "rpi.h"
#include "rpi-constants.h"
#include "rpi-interrupts.h"

#include "mmu.h"

fld_t * vm_ident_mmu_init(int start_p, void *int_vec);
void vm_ident_mmu_on(fld_t *pt);
void vm_ident_mmu_off(void);

// we use a non-zero domain id to test things out.
// (this shouldn't be a global) 
enum { dom_id = 1};

// kinda lame, but useful to have around.
enum { OneMB = 1024 * 1024 };

// used for tests.
extern volatile struct proc_state {
    fld_t *pt;
    unsigned sp_lowest_addr;
    unsigned fault_addr;
    unsigned dom_id;
    unsigned fault_count;

    // used for the die checks.
    unsigned die_addr;
} proc;



#endif
