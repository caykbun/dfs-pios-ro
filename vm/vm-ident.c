// What to do:
//   - replace the staff calls with calls to your routines in mmu.c
//
// section based: each section is 1MB.
//
// trivial setup of identity mappings where we map a virtual section 
// to the same physical section (e.g. 0x100000 maps to 0x100000)
//
// assumes a very simple address space:
//    - 1 section of code (starting at address 0x0)
//    - 1 section of user stack (ending at STACK_ADDR)
//    - 1 section of interrupt stack (ending at INT_STACK_ADDR)
//    - 1 section of heap (starting at one MB: 0x100000)
//    - 3 sections of GPIO ( 0x20000000 ---  0x20200000)
#include "vm/vm-ident.h"
#include "vector-base.h"

enum { MB = 1024 * 1024 };

// void mmu_reset(void);

// do one time initialization of MMU, setup identity mappings, and start
// running in mmu
fld_t * vm_ident_mmu_init(int start_p, void *int_vec) {
    // 1. init
    mmu_init();
    assert(!mmu_is_enabled());

    // ugly hack: we start by allocating a single page table at a known address.
    fld_t *pt = mmu_pt_alloc(4096);
    assert(pt == (void*)0x100000);

    // 2. setup mappings

    // map the first MB: shouldn't need more memory than this.
    mmu_map_section(pt, 0x0, 0x0, dom_id);
    // map the heap/page table: for lab cksums must be at 0x100000.
    mmu_map_section(pt, 0x100000,  0x100000, dom_id);

    mmu_map_section(pt, 0x20000000, 0x20000000, dom_id);
    mmu_map_section(pt, 0x20100000, 0x20100000, dom_id);
    mmu_map_section(pt, 0x20200000, 0x20200000, dom_id);
    
    mmu_map_section(pt, STACK_ADDR-MB, STACK_ADDR-MB, dom_id);
    mmu_map_section(pt, INT_STACK_ADDR-MB, INT_STACK_ADDR-MB, dom_id);

    // 3. install fault handler to catch if we make mistake.
    int_vec_reset(int_vec);

    // 4. start the context switch:

    // set up permissions for the one domain we use rn.
    domain_access_ctrl_set(0b01 << dom_id*2);

    if(start_p)
        vm_ident_mmu_on(pt);
    return pt;
}

void vm_ident_mmu_on(fld_t *pt) {
    // this code uses the sequence on B2-25
    set_procid_ttbr0(0x140e, dom_id, pt);

    // note: this routine has to flush I/D cache and TLB, BTB, prefetch buffer.
    mmu_enable();
    assert(mmu_is_enabled());
}

void vm_ident_mmu_off(void) {
    mmu_disable();
    assert(!mmu_is_enabled());
}
