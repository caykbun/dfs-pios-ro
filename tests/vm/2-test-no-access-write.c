// do a write with insufficient privleges and reset them.
#include "vm/vm-ident.h"
#include "libc/bit-support.h"

void vm_test(void) {
    extern uint32_t interrupt_vector_vm[];
    
    proc.sp_lowest_addr = STACK_ADDR - OneMB;
    proc.dom_id = dom_id;
    proc.pt = vm_ident_mmu_init(1, (void *)interrupt_vector_vm);
    assert(mmu_is_enabled());

    // 1: redo out of bounds stack write.
    // this is just to map it.
    volatile uint32_t *p = (void*)(STACK_ADDR - OneMB - 16);
    proc.fault_addr = (unsigned)p;
    *p = 11;
    assert(get32(p) == 11);

    // 2: do a write with insufficient privleges.
    printk("about to mark no access%x (vpn=%x)\n", p, mmu_sec_bits_only((uint32_t)p));
    uint32_t fa = proc.fault_addr = (unsigned)p;
    mmu_mark_sec_no_access(proc.pt, fa, 1);
    printk("HASH: marked no access%x (vpn=%x)\n", p, mmu_sec_bits_only(fa));
    assert(proc.fault_count == 1);
    put32(p, 12);
    assert(proc.fault_count == 2);
    assert(*p == 12);
    printk("HASH: done with invalid perm write to %x have=%d\n", p,*p);
    printk("HASH: test2: success!\n");

    // make sure the side-effect persists after
    vm_ident_mmu_off();
    assert(!mmu_is_enabled());
    assert(get32(p) == 12);
    printk("******** success ************\n");
}

void notmain() {
    mmu_be_quiet();
    kmalloc_init_set_start((void*)OneMB, OneMB);
    output("testing we can write and resume\n");
    check_vm_structs();
    vm_test();
}
