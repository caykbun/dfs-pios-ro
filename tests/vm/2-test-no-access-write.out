HASH: marked no access0x7effff0 (vpn=0x7e00000)
HASH: done with invalid perm write to 0x7effff0 have=12
HASH: test2: success!
